src/components/macro/macro.component.ts

```diff
@@ -35,65 +45,83 @@ export class MacroComponent implements OnInit, OnDestroy {
+            .do((macro: Macro) => {
+                if (macro) {
+                    this.currentId = macro.id;
+                }
+            })
+            .publishReplay(1);
+
+        this.macro$ = macroConnectable;
+        macroConnectable.connect();
     }
 
-    getMacro(id: number): Macro {
-        const config = this.uhkConfigurationService.getUhkConfiguration();
-        const macro: Macro = config.macros.elements.find(item => item.id === id);
-        if (macro) {
-            // Clone macro for editing
-            return new Macro().fromJsObject(macro.toJsObject());
+    showNewAction() {
+        if (this.activeEdit) {
+            this.hideActiveEditor();
         }
-        // @todo replace with notification
-        throw new Error('Macro not found');
-    }
 
-    addAction() {
-        this.hideOtherActionEditors(this.macro.macroActions.elements.length);
-        this.macro.macroActions.elements.push(undefined);
+        this.newMacro = undefined;
+        this.showNew = true;
     }
 
-    discardChanges() {
-        const id: number = this.macro.id;
-        this.macro = this.getMacro(id);
-        this.hasChanges = false;
+    hideNewAction() {
+        this.showNew = false;
     }
 
-    hideOtherActionEditors(index: number) {
-        this.macroItems.toArray().forEach((macroItem: MacroItemComponent, idx: number) => {
-            if (idx !== index) {
-                macroItem.cancelEdit();
-            }
-        });
+    addNewAction(macroAction: MacroAction) {
+        this.store.dispatch(MacroActions.addMacroAction(this.currentId, macroAction));
+        this.newMacro = undefined;
+        this.showNew = false;
     }
 
-    onEditAction(index: number) {
+    editAction(index: number) {
         // Hide other editors when clicking edit button of a macro action
-        this.hideOtherActionEditors(index);
+        if (this.activeEdit) {
+            this.hideActiveEditor();
+        }
+
+        this.activeEdit = index;
     }
 
-    onSaveAction(macroAction: MacroAction, index: number) {
-        this.hasChanges = true;
-        this.macro.macroActions.elements[index] = macroAction;
+    cancelAction() {
+        this.activeEdit = undefined;
     }
 
-    onDeleteAction(index: number) {
-        // @ todo show confirm action dialog
-        this.macro.macroActions.elements.splice(index, 1);
-        this.hasChanges = true;
+    saveAction(macroAction: MacroAction, index: number) {
+        this.store.dispatch(MacroActions.saveMacroAction(this.currentId, index, macroAction));
+        this.hideActiveEditor();
     }
 
-    ngOnDestroy() {
-        this.routeSubscription.unsubscribe();
+    deleteAction(macroAction: MacroAction, index: number) {
+        this.store.dispatch(MacroActions.deleteMacroAction(this.currentId, index, macroAction));
+        this.hideActiveEditor();
     }
 
+    private hideActiveEditor() {
+        this.macroItems.toArray().forEach((macroItem: MacroItemComponent, idx: number) => {
```

Comment:
I will move this if into this function, so that we can remove it on other places.